## Load/import packages

In [1]:
import json
import cv2
import numpy as np
import matplotlib.pyplot as plt

from os import listdir, mkdir
from os.path import splitext
from os.path import join

%matplotlib inline

## Pre-processing
1. Apply illumination normalization (histogram equalization)
2. Pose normalization
3. Store the image in a named folder same name as video. The image should be stored under the as videoname_frameXXXX ✔
4. Data augmentation to upsample some classes

## Directories
### Aff-Wild2

In [2]:
from load_filenames import (AW2_cropped_aligned_dir, 
                            AW2_cropped_aligned_folders, 
                            AW2_train_FN_split,
                            AW2_val_FN_split)

In [3]:
AW2_cropped_aligned_dir

'D:\\Aff-Wild2 Dataset\\Aff-wild2\\Images\\cropped_aligned'

In [4]:
AW2_cropped_aligned_folders

['1-30-1280x720',
 '10-60-1280x720',
 '10-60-1280x720_right',
 '100-29-1080x1920',
 '101-30-1080x1920',
 '102',
 '102-30-640x360',
 '103',
 '103-30-384x480',
 '104-17-720x480',
 '105',
 '105-30-1280x720',
 '106',
 '106-30-720x1280',
 '107',
 '107-30-640x480',
 '108',
 '108-15-640x480',
 '109-30-1280x720',
 '11-24-1920x1080',
 '110',
 '110-30-270x480',
 '111',
 '111-25-1920x1080',
 '112',
 '112-30-640x360',
 '113',
 '113-60-1280x720',
 '114',
 '114-30-1280x720',
 '115-30-1280x720',
 '116',
 '116-30-1280x720',
 '117',
 '117-25-1920x1080',
 '118',
 '118-30-640x480',
 '119',
 '119-30-848x480',
 '12-24-1920x1080',
 '120',
 '120-30-1280x720',
 '121',
 '121-24-1920x1080',
 '122',
 '122-60-1920x1080-1',
 '122-60-1920x1080-2',
 '122-60-1920x1080-3',
 '122-60-1920x1080-4',
 '122-60-1920x1080-5',
 '123',
 '123-25-1920x1080',
 '124-30-720x1280',
 '125',
 '125-25-1280x720',
 '126',
 '126-30-1080x1920',
 '127',
 '127-30-1280x720',
 '128',
 '128-24-1920x1080',
 '129',
 '129-24-1280x720',
 '13-30-1920

# Start Pre-processing

Each image has the shape (112, 112, 3)

To illumination normalize it, the image is converted to grayscale and gets the shape (112, 112)

In [26]:
for folder in AW2_cropped_aligned_folders[295:]:
    print(folder)

339
34-25-1920x1080
341
344
345
346
347
348
349
35-30-1920x1080
350
353
354
355
357
358
359
36-24-1280x720
360
361
362
363
364
365
367
368
369
37-30-1280x720
370
371
372
373
374
375
376
377
378
38-30-1920x1080
380
381
382
383
384
385
386
387
388
389
39-25-424x240
391
392
393
394
395
398
399
4-30-1920x1080
40-30-1280x720
400
402
403
406
407
408
409
41-24-1280x720
412
415
416
418
419
42-30-480x480
420
421
423
424
425
426
427
428
429
43-30-406x720
430
433
434
435
439
44-25-426x240
440
441
446
447
448
449
45-24-1280x720
450
46-30-484x360_left
46-30-484x360_right
47-30-654x480
48-30-720x1280
49-30-1280x720_left
49-30-1280x720_right
5-60-1920x1080-1
5-60-1920x1080-2
5-60-1920x1080-3
5-60-1920x1080-4
50-30-1920x1080
51-30-1280x720
52-30-1280x720_left
52-30-1280x720_right
53-30-360x480
54-30-1080x1920
55-25-1280x720
56-30-1080x1920
57-25-426x240
58-30-640x480
59-30-1280x720
6-30-1920x1080_left
6-30-1920x1080_right
60-30-1920x1080
61-24-1920x1080
62-30-654x480
63-30-1920x1080
64-24-640x360
65-3

In [5]:
from skimage import io, color, exposure
import time

In [27]:
AW2_normalized_dir = r"D:\Aff-Wild2 Dataset\Aff-wild2\Images\cropped_aligned_normalized"

def pre_processing():
    start = time.time()
    for folder in AW2_cropped_aligned_folders[295:]:
        folder_dir = join(AW2_cropped_aligned_dir,folder)
        frames_FN = listdir(folder_dir)
        print(f"Done! Time ran since start: {round(time.time()-start)//60 }:{round(time.time()-start)%60}")
        print(folder)
        # Create a directory with the name of the video to save all the normalized frames
        try:
            mkdir(join(AW2_normalized_dir, folder))
        except FileExistsError:
            print("Directory already exists!")

        for frame_name in frames_FN:
            if frame_name == ".DS_Store":
                continue
            frame = io.imread(join(folder_dir, frame_name))
            frame_g = color.rgb2gray(frame)
            frame_g_hist = exposure.equalize_adapthist(frame_g)
            frame_rgb_hist = color.gray2rgb(frame_g_hist)
            
            file = join(join(AW2_normalized_dir, folder), frame_name)
            io.imsave(file, (frame_rgb_hist*255).astype(np.uint8))


In [28]:
pre_processing()

Done! Time ran since start: 0:0
339
Directory already exists!
Done! Time ran since start: 0:30
34-25-1920x1080
Done! Time ran since start: 1:28
341
Done! Time ran since start: 2:0
344
Done! Time ran since start: 2:37
345
Done! Time ran since start: 3:0
346
Done! Time ran since start: 3:58
347
Done! Time ran since start: 4:27
348
Done! Time ran since start: 5:13
349
Done! Time ran since start: 5:52
35-30-1920x1080
Done! Time ran since start: 9:29
350
Done! Time ran since start: 9:51
353
Done! Time ran since start: 10:17
354
Done! Time ran since start: 11:12
355
Done! Time ran since start: 11:28
357
Done! Time ran since start: 11:41
358
Done! Time ran since start: 11:44
359
Done! Time ran since start: 12:21
36-24-1280x720
Done! Time ran since start: 13:12
360
Done! Time ran since start: 13:22
361
Done! Time ran since start: 14:7
362
Done! Time ran since start: 14:58
363
Done! Time ran since start: 15:30
364
Done! Time ran since start: 15:58
365
Done! Time ran since start: 16:8
367
Done! 

C:\Users\ion\anaconda3\envs\tf-gpu\lib\site-packages\ipykernel_launcher.py:25: UserWarning: D:\Aff-Wild2 Dataset\Aff-wild2\Images\cropped_aligned_normalized\video65\01640.jpg is a low contrast image
C:\Users\ion\anaconda3\envs\tf-gpu\lib\site-packages\ipykernel_launcher.py:25: UserWarning: D:\Aff-Wild2 Dataset\Aff-wild2\Images\cropped_aligned_normalized\video65\02498.jpg is a low contrast image
C:\Users\ion\anaconda3\envs\tf-gpu\lib\site-packages\ipykernel_launcher.py:25: UserWarning: D:\Aff-Wild2 Dataset\Aff-wild2\Images\cropped_aligned_normalized\video65\02499.jpg is a low contrast image


Done! Time ran since start: 188:23
video66
Done! Time ran since start: 188:24
video67
Done! Time ran since start: 188:43
video69
Done! Time ran since start: 190:7
video7
Done! Time ran since start: 190:37
video70
Done! Time ran since start: 192:33
video71
Done! Time ran since start: 194:25
video72
Done! Time ran since start: 194:26
video73
Done! Time ran since start: 195:27
video74_left
Done! Time ran since start: 198:45
video74_right
Done! Time ran since start: 202:11
video75
Done! Time ran since start: 202:17
video76
Done! Time ran since start: 203:1
video77
Done! Time ran since start: 203:34
video78
Done! Time ran since start: 203:38
video79
Done! Time ran since start: 204:19
video8
Done! Time ran since start: 205:21
video80
Done! Time ran since start: 205:26
video81
Done! Time ran since start: 206:51
video82
Done! Time ran since start: 208:43
video83
Done! Time ran since start: 209:10
video84
Done! Time ran since start: 209:16
video85
Done! Time ran since start: 210:41
video86_1
Do

*** 

## Move images in their correct train, val and test set directory

In [ ]:
import shutil

In [ ]:
# Load Aff-Wild2 AW2_train_classes.json
with open("AW2_train_classes.json", "r") as fp:
    AW2_train_classes = json.load(fp)
    
# Load Aff-Wild2 AW2_val_classes.json
with open("AW2_train_classes.json", "r") as fp:
    AW2_val_classes = json.load(fp)

In [ ]:
train_dir = r'D:\Aff-Wild2 Dataset\Aff-wild2\Sets\train'
val_dir = r'D:\Aff-Wild2 Dataset\Aff-wild2\Sets\val'
test_dir = r'D:\Aff-Wild2 Dataset\Aff-wild2\Sets\test'

In [ ]:
# Move training files
for folder, ext in AW2_train_FN_split:
    shutil.move(join(AW2_normalized_dir, folder), train_dir)
    
for folder, ext in AW2_val_FN_split:
    shutil.move(join(AW2_normalized_dir, folder), val_dir)

for folder in listdir(AW2_normalized_dir):
    shutil.move(join(AW2_normalized_dir, folder), test_dir)

In [ ]:
# Exctract image from each videofolder and move it to the corresponding class folder
train_perclass_dir = r'D:\Aff-Wild2 Dataset\Aff-wild2\Sets_per_class\train'
label_folders = listdir(train_perclass_dir)

for folder in listdir(train_dir)[:2]:
    ##e.g. returns folder: 51-30-1280x720
    
    labels = AW2_train_classes.get(folder +'.txt')
    
    for file, label in zip(listdir(folder), labels)[:10]:
        if label in [0,1,2,3,4,5,6]:
            new_dir = join(train_perclass_dir, labelfolders[label])
        else:
            continue
        shutil.copy(join(folder,file), new_dir)
        os.rename(join(new_dir,file), join(new_dir, folder+file))
        